Create New Export: angular => knockout

orangejenny opened the pull request:
This moves the behavior of the create new export button & modal to knockout. It leaves the rest of the export list page using angular.

For review, it's probably best to start by skimming the comments/structure of [create_export.js](https://github.com/dimagi/commcare-hq/blob/e9d013fca35af4400d0d951093a8ce7fd1d59381/corehq/apps/export/static/export/js/create_export.js), then going back and reviewing by commit.

@nickpell / @esoergel 

[QA ticket]( https://manage.dimagi.com/default.asp?284748)

orangejenny committed:
Get rid of extra-large creation message

orangejenny committed:
Removed angular control from the list of exports

Comment out the export list markup that's being converted from angular to knockout.
Leave the 'Create New Export' modal under the control of angular.

orangejenny committed:
angular => knockout in export_list_controller.html

orangejenny committed:
angular => knockout in loading_exports.html

orangejenny committed:
angular => knockout in table header

orangejenny committed:
angular => knockout for table rows and a few easy table cells

orangejenny committed:
Support bulk export

orangejenny committed:
Support copy

orangejenny committed:
Support enable/disable button (not enabling/disabling)

orangejenny committed:
Support Export and Edit Filters buttons (but not Edit Filters modal)

orangejenny committed:
Support Edit button

orangejenny committed:
Add name, description, and warnings

orangejenny committed:
Kill remaining djangular template tags

orangejenny committed:
Support daily saved export size and last updated stats

orangejenny committed:
Support update and copy buttons

orangejenny committed:
Support warning about no data

orangejenny committed:
Modal to update data

orangejenny committed:
Auto update modal

orangejenny committed:
Move analytics to knockout and remove some now unused angular

orangejenny committed:
Split out views for enabling/disabling and updating

orangejenny committed:
Move polling logic to knockout

orangejenny committed:
Tighten permissions on ajax views

orangejenny committed:
Drop .bulk-checkbox styles, which aren't having an effect

orangejenny committed:
Support modal to edit filters (w=1)

orangejenny committed:
Delete some no-longer-used angular

Delete angular-specific tests

It's a shame to be deleting tests, but they aren't especially reusable with the new knockout code.

orangejenny committed:
Not all exports have a formname (case exports)

orangejenny committed:
Fixed usage of page-level modelType and isDeid options

orangejenny committed:
Moved polling from exportListModel to exportModel

orangejenny committed:
Refactored to limit angular control.

Take the 'Create New' button out of angular's jurisdiction.
Add a skeletal knockout model for creating a new export.
Leave angular in charge of the list of exports.

orangejenny committed:
Added modal header and footer data bindings

orangejenny committed:
Added data bindings for all .modal-body elements except the main one

orangejenny committed:
Added data bindings for everything in the main .modal-body except the django form

orangejenny committed:
Added django form

- Split out get_app_data_drilldown_values as its own view
- Get rid of "try 3 times" logic, which is bespoke and arbitrary

orangejenny committed:
Added drilldown interaction

orangejenny committed:
Added support for disabling create button until form is valid

orangejenny committed:
Added support for resetting form

orangejenny committed:
Added support for form submission

orangejenny committed:
Removed references to hq.app_data_drilldown

orangejenny committed:
Cleaned up error handling

orangejenny committed:
Added notion of strictness to initial page data.

Ideally initial page data would be strict by default, but there's some amount of code
using it that relies on non-strictness.

orangejenny committed:
Removed angular tests.

Most of the tests just check that the js variables are set to the values
passed back by the initial ajax call. It would be good to have tests of
the drilldown interactivity, but these aren't that, and creating those
right now is prohibitively time-consuming because the drilldown is
based on a django form.

orangejenny committed:
Removed support for restoring deleted apps.

This was implemented in the original exports page rewrite,
https://github.com/dimagi/commcare-hq/pull/8399/
The "Deleted" option was then removed from the app type dropdown in
https://github.com/dimagi/commcare-hq/pull/10663 but the UI for restoring
deleted apps was not removed.

orangejenny committed:
Renamed export_list.js => export_list_main.js

orangejenny committed:
Moved createExportModel into its own file.

Useful because export_list_main.js is also going to eventually contain
knockout models for the export list part of this page.

orangejenny committed:
Removed createExportModel's dependency on initialPageData.

orangejenny committed:
Added documentation, renamed variables, and moved code blocks.

The diff on this one is messy. These are fairly mechanical changes: renaming variables
(mostly replacing 'form' with 'drilldown' since there are several meanings of 'form' in
this code) and then reordering code blocks to group related functionality together.

orangejenny committed:
Removed form_placeholders param

Removed case_placeholders param

Removed form_labels param

orangejenny committed:
Linted

orangejenny committed:
Moved exportListModel and exportModel to a separate file.

nickpell commented on corehq/apps/hqwebapp/static/hqwebapp/js/initial_page_data.js:
```diff
@@ -39,10 +39,13 @@ hqDefine('hqwebapp/js/initial_page_data', ['jquery', 'underscore'], function ($,
         if (_initData[name] === undefined) {
             _initData = gather(data_selector, _initData);
         }
+        if (strict && _initData[name] === undefined) {
```
Thoughts on using ```in``` instead of checking for ```undefined```?

https://stackoverflow.com/questions/1098040/checking-if-a-key-exists-in-a-javascript-object?answertab=votes#tab-top

The current version is probably ok for this code, but it seems like a pattern we shouldn't encourage.

orangejenny replied:
Good point - 3a69de20c9

nickpell replied:
👍

nickpell commented on corehq/apps/export/templates/export/partials/export_list_create_export_modal.html:
```diff
@@ -136,36 +83,28 @@ <h4>
                     {% blocktrans %}
                         It seems as though you haven't submitted any data to CommCare HQ.
-                        Please deploy an Application and submit data before creating any {{ export_type_plural }}.
+                        Please deploy an application and submit data before creating any {{ export_type_plural }}.
```
yay capitalization

nickpell commented on corehq/apps/app_manager/fields.py:
```diff
@@ -165,14 +165,9 @@ def get_app_sources(domain):
-    Holding off to a different PR, as I want to isolate QA to just the exports
-    (the first thing to use this) --Biyeun
-
+    """
```
👍

orangejenny replied:
The best of intentions...

nickpell commented on corehq/apps/export/static/export/js/create_export.js:
```diff
@@ -0,0 +1,311 @@
+        self._initSelect2 = function (observable, fieldSlug) {
+            return function (fieldData) {
+                if (fieldSlug) {
+                    observable('');
```
What is this?

orangejenny replied:
It clears the value of the field. I'm not 100% sure it's necessary, it's just the equivalent of [this line](https://github.com/dimagi/commcare-hq/pull/22128/files/4594a7c493720fc970a6da53be57b4c3e0bff56f#diff-70abf7231c07d7517f2229435dda567fL88).

nickpell replied:
Interesting.  I just can't imagine the new line having any side-effects (but I've been wrong before...)

nickpell reviewed (approved):
A few minor questions.  Good to see comments being updated.

orangejenny committed:
Added support for saving 'Edit Filters' form

orangejenny committed:
Moved filter form validation to knockout

orangejenny committed:
Replaced undefined check with _.has

orangejenny committed:
Added default values for a few attributes.

These will be turned into observables in knockout by fromJS,
so they need to have a default value.

orangejenny committed:
Moved if bindings to wrapper containers.

When an if binding is false, the element with the binding still exists,
only the markup inside it is missing. Using a visible binding instead
would hide these containers, but their descendents would still exist
and have their data bindings evaluated, so expressions like `emailedExport.hasFile()`
would error when there's no emailedExport.

orangejenny committed:
Replaced some if bindings with visible bindings.

For these ones, the entire element should disappear if the condition is false.
There aren't any problematic data bindings in their descendents.

orangejenny committed:
Fixed emwf values not displaying correctly

orangejenny committed:
Fixed lint errors

esoergel reviewed (approved):
exciting stuff

orangejenny commented on corehq/apps/export/views.py:
```diff
@@ -727,6 +719,27 @@ def get_saved_export_progress(self, in_data):
+    elif model_type == 'case':
+        response = rmi_helper.get_case_rmi_response()
+    else:
+        response = rmi_helper.get_dual_model_rmi_response()
```
haha, yeah, at one point while working on this I had ambitions of reworking `ApplicationDataRMIHelper`. I have now shelved those ambitions.

orangejenny commented on corehq/apps/export/static/export/js/export_list.js:
```diff
@@ -35,6 +35,29 @@ hqDefine("export/js/export_list", function () {
+            // TODO
+            return false;
+        };
+
```
yep, definitely agree with that principle - I think for this code in particular the TODO was a reminder to go find the analogous code later and fill in this function

orangejenny committed:
Merge branch 'master' into jls/ko-exports-create

orangejenny committed:
Added missing imports

orangejenny committed:
Moved generation of bulk export value to form submission

orangejenny committed:
Fixed lint error

orangejenny committed:
Removed app_manager#fields from Gruntfile

orangejenny committed:
Added .ko-template to #export-list to prevent flash on page load

orangejenny committed:
Merge pull request #22170 from dimagi/jls/ko-exports-list

List Exports: angular => knockout

orangejenny committed:
Merge branch 'master' into jls/ko-exports-create

orangejenny committed:
Fixed test_edit_daily_saved_export_filters
